sysboundary/maxwellian.cpp

```diff
@@ -0,0 +1,303 @@
+
+      std::vector<vmesh::GlobalID> blocksToInitialize =
+          this->findBlocksToInitialize(popID, templateCell, rho, T, Vx, Vy, Vz);
+      Realf* data = templateCell.get_data(popID);
+
+      for (vmesh::GlobalID i = 0; i < blocksToInitialize.size(); ++i) {
+         const vmesh::GlobalID blockGID = blocksToInitialize[i];
+         const vmesh::LocalID blockLID = templateCell.get_velocity_block_local_id(blockGID, popID);
+         const Real* block_parameters = templateCell.get_block_parameters(blockLID, popID);
+         creal vxBlock = block_parameters[BlockParams::VXCRD];
+         creal vyBlock = block_parameters[BlockParams::VYCRD];
+         creal vzBlock = block_parameters[BlockParams::VZCRD];
+         creal dvxCell = block_parameters[BlockParams::DVX];
+         creal dvyCell = block_parameters[BlockParams::DVY];
+         creal dvzCell = block_parameters[BlockParams::DVZ];
+
+         // Calculate volume average of distribution function for each cell in the block.
+         for (uint kc = 0; kc < WID; ++kc)
+            for (uint jc = 0; jc < WID; ++jc)
+               for (uint ic = 0; ic < WID; ++ic) {
+                  creal vxCell = vxBlock + ic * dvxCell;
+                  creal vyCell = vyBlock + jc * dvyCell;
+                  creal vzCell = vzBlock + kc * dvzCell;
+                  Real avr = 0.0;
+                  if (speciesParams[popID].nVelocitySamples > 1) {
+                     creal d_vx = dvxCell / (speciesParams[popID].nVelocitySamples - 1);
+                     creal d_vy = dvyCell / (speciesParams[popID].nVelocitySamples - 1);
+                     creal d_vz = dvzCell / (speciesParams[popID].nVelocitySamples - 1);
+                     for (uint vi = 0; vi < speciesParams[popID].nVelocitySamples; ++vi)
+                        for (uint vj = 0; vj < speciesParams[popID].nVelocitySamples; ++vj)
+                           for (uint vk = 0; vk < speciesParams[popID].nVelocitySamples; ++vk) {
+                              avr += maxwellianDistribution(popID, rho, T, vxCell + vi * d_vx - Vx,
+                                                            vyCell + vj * d_vy - Vy, vzCell + vk * d_vz - Vz);
+                           }
+
+                     avr /= speciesParams[popID].nVelocitySamples * speciesParams[popID].nVelocitySamples *
+                            speciesParams[popID].nVelocitySamples;
+                  } else {
+                     avr = maxwellianDistribution(popID, rho, T, vxCell + 0.5 * dvxCell - Vx,
+                                                  vyCell + 0.5 * dvyCell - Vy, vzCell + 0.5 * dvzCell - Vz);
+                  }
+
+                  if (avr != 0.0) {
+                     data[blockLID * WID3 + cellIndex(ic, jc, kc)] = avr;
+                  }
+               } // for-loop over cells in velocity block
+      }          // for-loop over velocity blocks
+
+      templateCell.get_population(popID).RHO_R = templateCell.get_population(popID).RHO;
```

Comment:
Maybe this should be done after `calculateCellMoments`? Otherwise it would get the RHO value from last templateCell?